Python BCs

chleh opened the pull request:
Python code for BCs
================
* [Hertz contact](https://github.com/chleh/ogs6/blob/python-bcs-for-pr/Tests/Data/Mechanics/Linear/Python/hertz-contact-bc.py) // [docu](https://deploy-preview-2170--ogs.netlify.com/docs/benchmarks/python-bc/hertz-contact/hertz-contact/)
* [Laplace equation with BCs by given functions](https://github.com/chleh/ogs6/blob/python-bcs-for-pr/Tests/Data/Elliptic/square_1x1_GroundWaterFlow_Python/bcs_laplace_eq.py) // [docu](https://deploy-preview-2170--ogs.netlify.com/docs/benchmarks/python-bc/python-laplace-eq/)

Limitations (not solved in this PR)
===========================
* Only works if all physical fields have the same ansatz functions, i.e., not for HM.
* The normal traction BC is special and not covered by this PR.
* Constraint boundary conditions (#2145) cope with secondary variables (fluxes). Fluxes cannot be *passed to* the Python scripts by the code of this PR.

TODO
=====
* [x] finish Hertz contact test
* [x] test with nonlinear BC
* [x] test with multi-variable process
* [x] input file documentation
* [x] maybe fix possible Python version mismatch (presumably between VTK and OGS)
* [x] fix git lfs tricks (*.pnK)
* [x] squash

Review
======
* [ ] attribute used
* [x] VTK Python version mismatch
* [x] discuss NLSolver changes.
* [ ] Large tests & Python?

Issues to open
============
* [x] Maybe the two versions of deriveBCMap could be merged into one?
* [x] Python BC for normal traction
* [x] Python BC for mixed FEM ansatz functions
* [x] passing secondary variables to Python BC

~~***Don't review yet! Soon: review commit-wise.***~~

chleh commented:
https://github.com/chleh/ogs6/blob/95e2e25c862e6979129e0143e36a7126886c2ed0/Tests/Data/Elliptic/square_1x1_GroundWaterFlow_Python/bcs_laplace_eq.py#L1-L48

endJunction commented:
I'm afraid to ask, but 132kloc?

chleh commented:
Yes. I copied OGS, and Python (2 and 3), and combined the two, and now there are Python BCs. :rofl: 
No. I assume its due to WIP ParaView state files. I'll remove them. Maybe I'll keep some of them to simplify the validation of (changed) reference solutions.

bilke commented:
@chleh Jenkins unstable (yellow) means there are Doxygen warnings:

https://jenkins.opengeosys.org/job/ufz/job/ogs/job/PR-2170/21/warnings18Result/

chleh commented:
@bilke, thanks for the hint. I almost started to be puzzled.

chleh commented:
Strange now windows compilation tells me
> Unknown compiler version - please run the configure tests and report the results

https://jenkins.opengeosys.org/blue/organizations/jenkins/ufz%2Fogs/detail/PR-2170/27/pipeline/46

bilke commented:
> Unknown compiler version - please run the configure tests and report the results

This comes from Boost, you have to ignore it...

endJunction commented on ProcessLib/CMakeLists.txt:
```diff
@@ -37,4 +37,14 @@ endif()
+        # snprintf might be defined even though it's not needed. Precompiled headers
+        # cause that this macro definition is visible in all of ProcessLib.
+        # That might lead to clashes with jsoncpp. In particular the following
+        # compilation error was observed on MSVC:
```
I could imagine that msvc version and python version would be helpful...

chleh replied:
Issue solved. Code and comment removed.

chleh commented:
Thanks for the fast response. Currently it seems like I cannot add any git lfs file. I tried with a fresh clone of my fork, added a png. Git push shows:
```
$ git push -u origin test-lfs 
Locking support detected on remote "origin". Consider enabling it with:
  $ git config lfs.https://gitlab.opengeosys.org/ogs/ogs.git/info/lfs.locksverify true
info: Uploading failed due to unsupported Content-Type header(s).                                              
info: Consider disabling Content-Type detection with:
info: 
info:   $ git config lfs.contenttype false
Uploading LFS objects:   0% (0/1), 130 KB | 0 B/s, done
...
```

Could anybody please check if you can add git LFS stuff to your own forks? Just check if it's my personal problem.

chleh commented:
Thanks for the fast response. Currently it seems like I cannot add any git lfs file. I tried with a fresh clone of my fork, added a png. Git push shows:
```
$ git push -u origin test-lfs 
Locking support detected on remote "origin". Consider enabling it with:
  $ git config lfs.https://gitlab.opengeosys.org/ogs/ogs.git/info/lfs.locksverify true
info: Uploading failed due to unsupported Content-Type header(s).                                              
info: Consider disabling Content-Type detection with:
info: 
info:   $ git config lfs.contenttype false
Uploading LFS objects:   0% (0/1), 130 KB | 0 B/s, done
...
```

Could anybody please check if you can add git LFS stuff to your own forks? Just check if it's my personal problem.

endJunction commented on ProcessLib/BoundaryCondition/PythonBoundaryCondition.cpp:
```diff
@@ -0,0 +1,225 @@
+                "script.");
+            return;
+        }
+        if (val.first)
```
if (!val,first) return

endJunction commented on ProcessLib/BoundaryCondition/PythonBoundaryCondition.cpp:
```diff
@@ -0,0 +1,225 @@
+        // gather primary variable values
+        primary_variables.clear();
+        auto const num_var = _dof_table_boundary->getNumberOfVariables();
+        for (int var = 0; var < num_var; ++var)
```
check for HM

endJunction commented on Applications/CLI/CMakeLists.txt:
```diff
@@ -5,6 +5,47 @@ target_link_libraries(ogs
+
+    add_library(ogs_python_bindings STATIC ogs_python_bindings.cpp)
+    target_link_libraries(ogs_python_bindings PRIVATE pybind11::embed)
+    target_include_directories(ogs_python_bindings PRIVATE ${PYTHON_INCLUDE_DIRS})
```
necessary?

chleh replied:
Cleaned up.

endJunction commented on Applications/CLI/CMakeLists.txt:
```diff
@@ -5,6 +5,47 @@ target_link_libraries(ogs
+    target_link_libraries(ogs PRIVATE pybind11::embed)
+
+    add_library(ogs_python_bindings STATIC ogs_python_bindings.cpp)
+    target_link_libraries(ogs_python_bindings PRIVATE pybind11::embed)
```
Consider PUBLIC.

chleh replied:
Cleaned up.

endJunction commented on Applications/CLI/ogs_embedded_python.cpp:
```diff
@@ -0,0 +1,52 @@
+    // OpenGeoSys Python module. The name is generated by pybind11. If it is not
+    // obvious that this symbol is actually used, the linker might remove it
+    // under certain circumstances.
+    mark_used(&pybind11_init_impl_OpenGeoSys);
```
just an idea...
Maybe declaring the `pybind11_init_impl_OpenGeoSys` function with `__attribute__((used))` helps (at least for gcc and clang).

chleh replied:
`__attribute__(used)` is gcc-specific. I didn't find the same for MSVC:
https://msdn.microsoft.com/de-de/library/74f4886t.aspx

Maybe `__declspec(dllexport)` might help?
https://msdn.microsoft.com/de-de/library/3y1sfaz2.aspx
But is that also applicable for static libs?

Anyway, we'd have to distinguish between different compilers, which is also kind of hackish...

endJunction commented on NumLib/DOF/LocalToGlobalIndexMap.cpp:
```diff
@@ -249,7 +250,42 @@ LocalToGlobalIndexMap* LocalToGlobalIndexMap::deriveBoundaryConstrainedMap(
+
+    for (int i = 0; i < getNumberOfComponents(); ++i)
+    {
+        global_component_ids.push_back(i);
```
`std::iota(begin(global_component_ids), end(global_component_ids), getNumberOfComponents())`

chleh replied:
Indeed, verbose :smile:

endJunction commented on NumLib/DOF/LocalToGlobalIndexMap.cpp:
```diff
@@ -249,7 +250,42 @@ LocalToGlobalIndexMap* LocalToGlobalIndexMap::deriveBoundaryConstrainedMap(
+}
+
+std::unique_ptr<LocalToGlobalIndexMap>
+LocalToGlobalIndexMap::deriveBoundaryConstrainedMap(
```
Maybe the two versions of deriveBCMap could be merged into one?

chleh replied:
I'll open an issue.

endJunction commented on ProcessLib/BoundaryCondition/Python/PythonBoundaryConditionLocalAssembler.h:
```diff
@@ -0,0 +1,198 @@
+                "bulk_node_ids");
+
+        // gather primary variables
+        Eigen::MatrixXd primary_variables_mat(num_nodes, num_comp_total);
```
Optionally use compile time size for nodes' size...

chleh replied:
Hm... I think, I'd like to leave it all dynamic:
`Eigen::Matrix<double, ShapeFunction::size /*?*/, Eigen::Dynamic>` doesn't really look more pleasant.

endJunction commented on Tests/Data/Mechanics/Linear/Python/gen-unit-circle.py:
```diff
@@ -0,0 +1,154 @@
+    # self.GetOutputDataObject(0).GetPointData().AddArray(r2)
+    # self.GetOutputDataObject(0).GetPointData().AddArray(c2)
+
+    # self.GetOutputDataObject(0).SetPoints(pts) #.SetData(c2)
```
Maybe some debug comments could be removed...

chleh replied:
removed by now.

bilke commented on ProcessLib/BoundaryCondition/Python/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+    PUBLIC BaseLib MathLib MeshLib NumLib logog)
+
+target_link_libraries(ProcessLibBoundaryConditionPython
+    PRIVATE pybind11::pybind11)
```
You can merge this command with the previous one:

```cmake
target_link_libraries(ProcessLibBoundaryConditionPython
    PUBLIC BaseLib MathLib MeshLib NumLib logog
    PRIVATE pybind11::pybind11)
```

chleh replied:
All pvsm files will be deleted.
Thanks for the hint. I'll change it.

TomFischer commented on ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp:
```diff
@@ -0,0 +1,239 @@
+                                      MeshLib::MeshItemType::Node,
+                                      bulk_node_id};
+                auto const dof_idx =
+                    _bc_data.dof_table_bulk.getGlobalIndex(loc, var, comp);
```
Is it possible to compute the `dof_idx` also from the boundary dof table?

chleh replied:
I think, it's not. I think I tried, but the dof table for the boundary is for a hypothetical equation system that is assembled solely on the boundary (i.e., with no connection to the bulk). Therefore all boundary dof indices would be wrong. @endJunction, is that right? If unsure, I can also check that.

TomFischer commented on Tests/Data/Mechanics/Linear/Python/gen-unit-circle.py:
```diff
@@ -0,0 +1,153 @@
+a = np.empty(coords.shape[0])
+
+for i, (x, y, z) in enumerate(coords):
+    r = np.sqrt(x**2 + y**2)
```
Where is `r` used?

chleh replied:
In some old version of this script, long, long time ago. I'll remove it.

chleh commented on ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp:
```diff
@@ -0,0 +1,239 @@
+        if (!_flush)
+            return;
+
+        using namespace pybind11::literals;
```
For the line below, e.g. `"end"_a` is the equivalent to a Python named argument.

chleh commented on ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp:
```diff
@@ -0,0 +1,239 @@
+            return;
+
+        using namespace pybind11::literals;
+        pybind11::print("end"_a = "", "flush"_a = true);
```
It's equivalent to (Python) `print(end="", flush=True)`, i.e., no end-of-line will be printed and output will be flushed.
So the behaviour is described in the docu of he destructor. Is that sufficient?

chleh replied:
Yes. `_a` is a [user-defined literal](https://en.cppreference.com/w/cpp/language/user_literal).
There are also other such literals, e.g. `s` in C++14.
E.g. in
```
using namespace std::string_literals;
auto str = "test"s;
```
`str` is a `std::string`, not a character array.

chleh commented on web/content/docs/benchmarks/python-bc/hertz-contact/hertz-contact.md:
```diff
@@ -0,0 +1,65 @@
+values in every load step.
+Due to symmetry reasons a flat circular contact area of radius $a$ forms.
+
+{{< img src="../hertz-contact.pnK" >}}
```
The file names now are fixed.

chleh commented on ProcessLib/BoundaryCondition/Python/PythonBoundaryConditionLocalAssembler.h:
```diff
@@ -0,0 +1,198 @@
+                for (unsigned element_node_id = 0; element_node_id < num_nodes;
+                     ++element_node_id)
+                {
+                    auto* node = _element.getNode(element_node_id);
```
changed.

chleh commented on ProcessLib/BoundaryCondition/Python/PythonBoundaryConditionLocalAssembler.h:
```diff
@@ -0,0 +1,198 @@
+                        _data.dof_table_bulk.getGlobalIndex(loc, var, comp);
+                    if (dof_idx == NumLib::MeshComponentMap::nop)
+                    {
+                        // TODO extend Python BC to mixed FEM
```
`// TODO extend Python BC to mixed FEM ansatz functions` still fits on a single line (and therefore is displayed completely in QtCreator's todo list). I'd suggest tht wording.

chleh commented:
@bilke, thanks for the VTK Python version hint.
`Modules/vtkPython.cmake` contains some info **(1)**:
```
set(VTK_PYTHONPATH "/usr/lib/python2.7/site-packages")
set(vtkPython_LIBRARIES "/usr/lib/libpython2.7.so")
```
Furthermore `/usr/lib/cmake/vtk-8.1/vtkModuleMacrosPython.cmake`
contains `find_package(PythonInterp ${VTK_PYTHON_VERSION} REQUIRED)`.
But probably that won't help us, because it just searches for some locally installed Python intepreter.

And this information is not propagated to OGS. My `CMakeCache.txt` contains basically only **(2)**
`VTK_DIR:PATH=/usr/lib64/cmake/vtk-8.1` as interesting VTK related information.

Taking (1) and (2) together, one could probably detect the Python version VTK is built with, if any. But is that really stable/portable?

bilke commented on ThirdParty/CMakeLists.txt:
```diff
@@ -47,6 +47,33 @@ endif()
+        endif()
+    endfunction()
+
+    check_python_version_compatibility()
```
Looks good! If it works for you then I am fine. Jenkins will not complain as there are no VTK with Python binding packages used.

chleh commented:
I just rebased and squashed this PR. Thank you for the quick reviewing process and for your help.
Lately, netlify failed because hugo could not resolve some page reference. Locally, at my machine it works, however. @bilke, could you please have a look.

Unfortunately I had to exclude the Hertz contact test, because it requires a non-trivial change to the non-linear solver, cf. #2184.

Remaining for this PR are:
* I decided not to use `__attribute__(used)` because it's not portable and I don't know of a version that works for MSVC.
* The Elder test case is a LARGE one. Are LARGE tests tested with a Python build?

If this PR builds successfully, from my perspective it's ready to be merged if you agree.

endJunction commented:
LARGE tests are not tested.

bilke commented:
For the failing web build merge this: https://github.com/chleh/ogs6/pull/10
